text/0000-dyn-type.md

```diff
@@ -0,0 +1,2261 @@
+changing from a normal trait + `impl Trait for ..` to a dedicated syntax `auto trait Trait {}`.
+
+Our custom DST should provide their custom metadata, and the underlying DST struct representation.
+Therefore we choose our first syntax like this:
+
+```rust ,ignore
+dyn type CStr(()) = [c_char];
+//            ^~    ^~~~~~~~
+//             |     underlying representation
+//             |
+//             metadata
+
+dyn type Mat<T>(MatMeta) = [T];
+//              ^~~~~~~    ^~~
+//                    |     underlying representation
+//                    |
+//                    metadata
+```
+
+“`dyn`” is picked because it means dynamic,
+and recently becomes a keyword thanks to `dyn Trait` ([RFC #2113]),
+so we don’t need to introduce another unconstrained contextual keyword.
+Adding a `type` makes it spell out “dynamic type”, as well as preventing potential misinterpretation as a `dyn Trait`.
+
+### Customizing size and alignment
+
+In the most flexible sense, the size and alignment of custom DSTs are arbitrary functions and will be provided by user.
+
+```rust ,ignore
+// not part of this RFC
+trait CustomDst {
+    fn size_of_val(&self) -> usize;
+    fn align_of_val(&self) -> usize;
+}
+```
+
+Now, while `std::mem::{size_of_val, align_of_val}` takes the whole DST reference as input,
+it may not make sense for `align_of_val` to use its pointer as input. To wit, consider an unsized struct:
+
+```rust ,ignore
+struct S<T: ?Sized> {
+    a: u8,
+    b: T,
+}
+let s: &S<Dst> = ...;
+let t: &Dst = &s.b;
+```
+
+In order to get the address `&s.b`, we first need to find out the offset of `b` in `S<Dst>`.
+This requires knowing the alignment of `Dst`, which is `align_of_val(&s.b)`, and that leads to a circular dependency!
+Hence, our `align_of_val` should not depend on the pointer part of the reference at all.
+
+```rust ,ignore
+// not part of this RFC
+trait CustomDst {
+    fn size_of_val(val: *const Self) -> usize;
+    fn align_of_meta(meta: Self::Metadata) -> usize;
+}
+```
+
+The alignment needs to read `Metadata` because of how trait objects are implemented.
+Consider the types `*S<u8>` and `*S<u16>`.
+
+```text
++=======+    +===============+
+| S<u8> |    |    S<u16>     |
++=======+    +===============+
+| a | b |    | a |   |   b   |
++---+---+    +---+   +-------+
+0   1   2    0   1   2   3   4
+```
+
+Both types can be *coerced* into `*S<dyn Debug>`.
+The type `S<dyn Debug>` itself does not tell us the alignment of field `b`.
+This can only be obtained via the type information stored in the trait object’s metadata.
+
+### Reduction and regular/inline DST
+
+We want to be able to treat every aspect of the a DST as if the ordinary DST to maintain safety.
+So what the user should provide is not `size_of_val`/`align_of_meta`,
+but a function which produces the original DST which we compute the size and alignment from it instead.
+We call such process *reduction*, and the custom DST is *derived* from that ordinary DST.
```

Comment:
Should probably disambiguate "reduction" with "DST-reduction" or something since reduction is a more general concept (evaluation).